src/plugins/data/public/ui/query_editor/query_editor.tsx

```diff
@@ -240,11 +240,10 @@ export default class QueryEditorUI extends Component<Props, State> {
 
     if (!isEqual(this.props.query.dataset, prevQuery.dataset)) {
       if (this.inputRef) {
-        const newQuery = this.queryString.getInitialQuery();
-        const newQueryString = newQuery.query;
-        if (this.inputRef.getValue() !== newQueryString) {
+        const newQueryString = this.props.query.query;
```

Comment:
i think now when i select the dataset for the new language the query string is not updated with the updated dataset.

Reply:
recreation steps:
* have an index pattern and navigate to discover. load results 
* select an index from a cluster.
* notice how the query (if for sql and ppl) still references the old dataset